icu4c/source/common/unistr.cpp

```diff
@@ -1563,6 +1563,11 @@ UnicodeString::doAppend(const UChar *srcChars, int32_t srcStart, int32_t srcLeng
   }
 
   int32_t oldLength = length();
+  if (INT32_MAX - oldLength < srcLength) {
```

Comment:
An alternative would be to use uprv_add32_overflow.  However, what you have written here is likely faster since we know the inputs are positive.

FYI for the future:

```cpp
int32_t oldLength = length();
int32_t newLength;
bool overflow = uprv_add32_overflow(oldLength, srcLength, &newLength);
if (overflow) {
  setToBogus();
  return *this;
}
```

Reply:
done